### PR TITLE
Edge case: cross-shard scheduled miniblocks with construction state = final.

### PR DESCRIPTION
**Pull request by andreibancioiu**
- Handle edge case: cross-shard scheduled miniblocks with `construction state = final`.
 - Added an extra condition to ignore empty (and artificial) invalid miniblocks (does not affect flow)
 - Adjust docker images to set `StartInEpoch = false`
 - Update readme (e.g. on how to avoid Docker caching when building images)

### CODE REVIEW

File: server/provider/scheduledMiniblocks.go [hunk 1]
```diff
@@ -20,6 +20,13 @@ func (provider *networkProvider) simplifyBlockWithScheduledTransactions(block *d
 		return err
 	}
 
+	doSimplifyBlockWithScheduledTransactions(previousBlock, block, nextBlock)
+
+	return nil
+}
+
+// TODO after review: Move below (diff is easier to follow as it is now)
```
Comment (andreibancioiu) [inline server/provider/scheduledMiniblocks.go:28]: Will be removed after review.

File: server/provider/scheduledMiniblocks.go [hunk 2]
```diff
@@ -56,13 +64,16 @@ func hasOnlyNormalMiniblocks(block *data.Block) bool {
...
 	})
 }
 
 func removeScheduledMiniblocks(block *data.Block) {
 	removeMiniblocksFromBlock(block, func(miniblock *data.MiniBlock) bool {
-		return miniblock.ProcessingType == string(Scheduled)
+		hasProcessingTypeScheduled := miniblock.ProcessingType == dataBlock.Scheduled.String()
+		hasConstructionStateNotFinal := miniblock.ConstructionState != dataBlock.Final.String()
+
+		return hasProcessingTypeScheduled && hasConstructionStateNotFinal
```
Comment (bogdan-rosianu) [inline server/provider/scheduledMiniblocks.go:76]: might extract a new variable called `shouldRemove` which will be equal to `hasProcessingTypeScheduled && hasConstructionStateNotFinal`

File: server/provider/scheduledMiniblocks.go [hunk 3]
```diff
@@ -96,7 +107,10 @@ func findScheduledTransactionsHashes(block *data.Block) map[string]struct{} {
 	invalidTxs := make(map[string]struct{})
 
 	for _, miniblock := range block.MiniBlocks {
-		if miniblock.ProcessingType == string(Scheduled) {
+		hasProcessingTypeScheduled := miniblock.ProcessingType == dataBlock.Scheduled.String()
+		hasConstructionStateNotFinal := miniblock.ConstructionState != dataBlock.Final.String()
+
+		if hasProcessingTypeScheduled && hasConstructionStateNotFinal {
```
Comment (bogdan-rosianu) [inline server/provider/scheduledMiniblocks.go:113]: might extract a new variable called `shouldAddToInvalidTxs`

File: server/provider/scheduledMiniblocks.go (hunk 2 above)
Comment (andreibancioiu) [reply server/provider/scheduledMiniblocks.go]: Fixed.

File: server/provider/scheduledMiniblocks.go (hunk 3 above)
Comment (andreibancioiu) [reply server/provider/scheduledMiniblocks.go]: Fixed & renamed (the map previously had a wrong name).

**Review by bogdan-rosianu — APPROVED** [review]

File: server/provider/scheduledMiniblocks.go [hunk 4]
```diff
@@ -56,13 +64,16 @@ func hasOnlyNormalMiniblocks(block *data.Block) bool {
...
+		return miniblock.ProcessingType == dataBlock.Processed.String()
 	})
 }
 
 func removeScheduledMiniblocks(block *data.Block) {
 	removeMiniblocksFromBlock(block, func(miniblock *data.MiniBlock) bool {
-		return miniblock.ProcessingType == string(Scheduled)
+		hasProcessingTypeScheduled := miniblock.ProcessingType == dataBlock.Scheduled.String()
+		hasConstructionStateNotFinal := miniblock.ConstructionState != dataBlock.Final.String()
+		shouldRemove := hasProcessingTypeScheduled && hasConstructionStateNotFinal
```
Comment (AdoAdoAdo) [inline server/provider/scheduledMiniblocks.go:75]: :+1:

**Review by AdoAdoAdo — DISMISSED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**